src/MallardMessageHandlers/AuthenticationToken/AuthenticationTokenHandler.cs

```diff
@@ -149,11 +149,27 @@ TAuthenticationToken unauthorizedToken
 
 			_logger.LogDebug($"Requesting refresh for token: '{unauthorizedToken}'.");
 
+			// Wait for the semaphore.
 			await _semaphore.WaitAsync(ct);
 
 			// From this moment, the operation cannot be cancelled.
-			ct = CancellationToken.None;
+			var refreshedToken = await GetRefreshedAuthenticationToken(
+				CancellationToken.None,
+				request,
+				unauthorizedToken
+			);
 
+			// Release the semaphore.
+			_semaphore.Release();
```

Comment:
You should use a `try finally` to release the semaphore even when `GetRefreshedAuthenticationToken` fails.

Reply:
good call